packages/server/src/api/todo/adx/account/password-reset.ts

```diff
@@ -0,0 +1,105 @@
+import jwt from 'jsonwebtoken'
+import { ServerConfig } from '../../../../config'
+import { User } from '../../../../db/user'
+import { Server } from '../../../../lexicon'
+import * as locals from '../../../../locals'
+
+const PASSWORD_RESET_SCOPE = 'todo.adx.resetAccountPassword'
+
+export default function (server: Server) {
+  server.todo.adx.requestAccountPasswordReset(
+    async (_params, input, _req, res) => {
+      const { db, mailer, config } = locals.get(res)
+      const { email } = input.body
+
+      // @TODO should multiple users be able to have the same email?
+      const table = db.db.getRepository(User)
+      const users = await table.findBy({ email })
+
+      for (const user of users) {
+        // By signing with the password hash, this jwt becomes invalid once the user changes their password.
+        // This allows us to ensure it's essentially single-use (combined with the jwt's short-livedness).
+        const token = jwt.sign(
+          {
+            sub: user.did,
+            scope: PASSWORD_RESET_SCOPE,
+          },
+          getSigningKey(user, config),
+          { expiresIn: '15mins' },
+        )
+
+        await mailer.sendResetPassword({ token }, { to: user.email })
+      }
+
+      return {
+        encoding: 'application/json',
+        body: {},
+      }
+    },
+  )
+
+  server.todo.adx.resetAccountPassword(async (_params, input, _req, res) => {
+    const { db, config } = locals.get(res)
+    const { token, password } = input.body
+
+    const tokenBody = jwt.decode(token)
+    if (tokenBody === null || typeof tokenBody === 'string') {
+      return createInvalidTokenError('Malformed token')
+    }
+
+    const { sub: did, scope } = tokenBody
+    if (typeof did !== 'string' || scope !== PASSWORD_RESET_SCOPE) {
+      return createInvalidTokenError('Malformed token')
+    }
+
+    const table = db.db.getRepository(User)
+    const user = await table.findOneBy({ did })
+    if (!user) {
+      return createInvalidTokenError('Token could not be verified')
+    }
+
+    try {
+      jwt.verify(token, getSigningKey(user, config))
+    } catch (err) {
+      if (err instanceof jwt.TokenExpiredError) {
+        return createExpiredTokenError()
+      }
+      return createInvalidTokenError('Token could not be verified')
+    }
+
+    // Token had correct scope, was not expired, and referenced
+    // a user whose password has not changed since token issuance.
+
+    await db.updateUserPassword(did, password)
+
+    return {
+      encoding: 'application/json',
+      body: {},
+    }
+  })
+}
+
+const getSigningKey = (user: User, config: ServerConfig) =>
+  `${config.jwtSecret}::${user.password}`
+
+type ErrorResponse = {
```

Comment:
no need to do for this PR, but this type is relevant to other xrpc methods as well. & maybe we could go so far as to codegen error creators 👀

Reply:
Yeah, this was an interesting one.  The main reason I had to write this type was to make it easier to write the type `ErrorResponse & { error: 'SomeSpecificError' }` for those error creators.  Otherwise typescript wouldn't infer that the error creators were always returning an allowed `error` value for the handler.  On the upside, it means the handlers are typed specifically enough to know exactly the permitted errors per the schema :)

Reply:
(P.S. that would be pretty awesome!)